boat_simulator/common/sensors.py

```diff
@@ -60,98 +81,179 @@ def read(self, key: str) -> Any:
         # - wind vectors should add with noise vectors.
-        # Could consider using a __post_init__ function for this
+
         return (
             self._wind + self.wind_noisemaker.next()  # type: ignore
-            if self.wind_noisemaker is not None
+            if self.enable_noise
             else self._wind
         )
 
     @wind.setter
     def wind(self, wind: ScalarOrArray):
-        self._wind = wind
+
+        if not self.enable_delay:
+            self._wind = wind
+            return
+
+        if self.wind_queue_next:
+            self._wind = self.wind_next_value
+        else:
+            self.wind_queue_next = True
+
+        self.wind_next_value = wind
 
 
-@dataclass
 class GPS(Sensor):
     """
     Abstraction for GPS.
 
-    # TODO: Add delay functions.
-
     Properties:
         lat_lon (NDArray): Boat latitude and longitude (2x1 array)
         speed (Scalar): Boat speed
         heading (Scalar): Boat heading
-        lat_lon_noisemaker (Optional[GaussianGenerator | ConstantGenerator]):
-        Noise function to emulate sensor noise in latitude and longitude readings
-        speed_noisemaker (Optional[GaussianGenerator | ConstantGenerator]):
-        Noise function to emulate sensor noise in speed readings
-        heading_noisemaker (Optional[GaussianGenerator | ConstantGenerator]):
-        Noise function to emulate sensor noise in heading readings
+        enable_noise (bool): Enables noise for fields. False by default.
+        enable_delay (bool): Enables delay for fields. False by default.
     """
 
     lat_lon: NDArray
     speed: Scalar
     heading: Scalar
 
-    lat_lon_noisemaker: GPSGenerators = None
-    speed_noisemaker: GPSGenerators = None
-    heading_noisemaker: GPSGenerators = None
+    def __init__(
+        self,
+        lat_lon: NDArray,
+        speed: Scalar,
+        heading: Scalar,
+        stdev: Scalar = 1,
+        enable_noise: bool = False,
+        enable_delay: bool = False,
+    ):
+        super().__init__(enable_noise=enable_noise, enable_delay=enable_delay)
+        self._lat_lon = lat_lon
+        self._speed = speed
+        self._heading = heading
+
+        # TODO: Refactor the initialization of data fields and their respective delay controls.
+        # Warning: this is not easy!
+
+        # Delay Controls
+        self.lat_lon_queue_next: bool = False
+        self.lat_lon_next_value: NDArray = lat_lon
+
+        self.speed_queue_next: bool = False
+        self.speed_next_value: Scalar = speed
+
+        self.heading_queue_next: bool = False
+        self.heading_next_value: Scalar = heading
+
+        self.lat_lon_noisemaker: GaussianGenerator = GaussianGenerator(
+            mean=0, stdev=stdev
+        )
+        self.speed_noisemaker: GaussianGenerator = GaussianGenerator(
+            mean=0, stdev=stdev
+        )
+        self.heading_noisemaker: GaussianGenerator = GaussianGenerator(
+            mean=0, stdev=stdev
+        )
```

Comment:
The standard deviation will be at different units/scales for each of these properties. We should have three instead of one. We should also make it clear that the standard deviation is only used for noise generation in its variable name.

Reply:
Resolved.

Reply:
See https://github.com/UBCSailbot/sailbot_workspace/pull/311